### DISCUSSION TITLE
Замена значений по условию в rowList

### DISCUSSION THREAD

**Discussion by Vadimgrig2012** [totumonline/totum-ru-issues-and-discussions#316] [Как это сделать?] [answered]
Всем добрый день! 
Подскажите пожалуйста, каким образом можно передать значения полей в шаблон по условию? Объясню: есть крона, которая должна каждый день в 9 утра отправлять по почте пользователям некую напоминалку в виде таблицы через selectRowList по полям из таблицы циклов, где заявка = статусу "открыта". Шаблоны написаны, данные  отправляются, таблица прилетает. Все  ОК, но... Большинство полей в таблице циклов типа "дата" и если, к примеру, какое-то поле пустое, то форматированием задается значение "Пусто" и соответственно при передаче значения из этого поля в шаблон через selectRowList в ячейку шаблона не попадает вообще ничего и она остается пустой. Необходимо, что бы в нее прилетало например то же самое "Пусто".  Каким образом можно этого достичь?

Код выглядит вот таким образом:

```
=: if(condition: $sel != $#lc; then: $send)
	sel: selectList(table: 'zayavki'; field: 'id'; where: 'status' = "open")
	send: emailSend(to: $userssend; title: "Ежедневный отчет по открытым заявкам"; body: str`$body`)
		userssend: select(table: 'users'; field: 'email'; where: 'fio' = "ФИО")
		body: textByTemplate(template: "eml_sendnotif_kom"; data: $data)
		data: rowCreate(field: "table" = $row)
			row: rowCreate(field: "template" = "table_sendnotif_kom"; field: "data" = $rowlist)
				rowlist: selectRowList(table: 'zayavki'; sfields: json`["zakazchik","postavshchiki","predlojeniya"]`; field: 'nomer'; field: 'srok_podachi_zayavki_do'; field: 'zapros_napravlen'; field: 'predlojenie_zakazchiku'; where: 'status' = "open"; order: 'nomer' asc)
```

`srok_podachi_zayavki_do`, `zapros_napravlen`, `predlojenie_zakazchiku` - поля типа "дата" и при наличии в данных полях null необходимо, что бы подставлялось значение в шаблоне в ячейках "Пусто"/"Нет"/"---" (в общем что угодно, но не "").

Так же есть великая вероятность того, что в дальнейшем поля в таблицу циклов будут добавляться и типы полей будут разные и условия для подстановки значений соответственно тоже будут разные.

**Comment by ttm-alexeypolunin** [comment, marked as answer]
Вот небольшой код, который позволяет двумя вложенными `listReplace` перебрать любой `rowList` и по условиям заменить на  новые значения:

```
= : linkToDataJson(title: "*"; data: $data)

data: listReplace(list: $rowList; action: $rowReplace; value: "row")

	rowList: selectRowList(table: $#ntn; field: 'first_field'; field: 'second_field'; where: 'id' = json`[2,3,7,8]`)
	
		rowReplace: listReplace(list: $#row; action: $check_item; value: "item")
			check_item: if(condition: $#item = ""; then: "Not provided"; else: $#item)
```

Если нужно, что бы от названия поля зависел результат, то надо во вложенном listReplace использовать параметр `key` и использовать созданную им переменную в условиях:

```
= : linkToDataJson(title: "*"; data: $data)

data: listReplace(list: $rowList; action: $rowReplace; value: "row")

	rowList: selectRowList(table: $#ntn; field: 'first_field'; field: 'second_field'; where: 'id' = json`[2,3,7,8]`)
	
		rowReplace: listReplace(list: $#row; action: $check_item; key: "row_key"; value: "item")
			check_item: if(condition: $#item = ""; condition: $#row_key = "second_field";  then: "Not provided"; else: $#item)
```

**Reply by Vadimgrig2012** [reply]
Спасибо огромное! Оч помогли!
